Mathlib/RingTheory/OrderOfVanishing.lean

```diff
@@ -0,0 +1,221 @@
+-/
+lemma Ideal.quotOfMulSurjective {a : R} (I : Ideal R) :
+    Function.Surjective (Ideal.quotOfMul a I) := by
+  simp only [Ideal.quotOfMul]
+  exact Submodule.factor_surjective <|
+    Submodule.singleton_set_smul I a ▸ Ideal.smul_le_span {a} I
+/--
+The sequence `R ⧸ I →ₗ[R] R ⧸ (a • I) →ₗ[R] R ⧸ (Ideal.span {a})` given by multiplication
+by `a` then quotienting by the ideal generated by `a` is exact.
+-/
+lemma Ideal.quotMulExact {a : R} (I : Ideal R) :
+    Function.Exact (Ideal.mulQuot a I) (Ideal.quotOfMul a I) := by
+  simp only [exact_iff]
+  have : ker (Ideal.quotOfMul a I) = a • ⊤ := by
+    simp only [Ideal.quotOfMul, Submodule.factor, Submodule.mapQ, comp_id, Submodule.ker_liftQ,
+      Submodule.ker_mkQ]
+    suffices Submodule.map (Submodule.mkQ (a • I)) (Submodule.span R {a}) = a • ⊤ by exact this
+    rw [Submodule.map_span]
+    simp only [Submodule.mkQ_apply, Quotient.mk_eq_mk, Set.image_singleton, Ideal.Quotient.smul_top]
+  simp [this, Ideal.mulQuot, Submodule.mapQ.eq_1, Submodule.range_liftQ,
+   range_comp, Ideal.Quotient.smul_top, ← Ideal.submodule_span_eq, LinearMap.map_span]
+
+variable (R)
+/--
+The order of vanishing of `a * b` is the order of vanishing of `a` plus the order
+of vanishing of `b`.
+-/
+theorem CommRing.ord_mul {a b : R} (hb : b ∈ nonZeroDivisors R) :
+    Ring.ord R (a*b) = Ring.ord R a + Ring.ord R b := by
+  have :=  Module.length_eq_add_of_exact (Ideal.mulQuot b (Ideal.span {a}))
+          (Ideal.quotOfMul b (Ideal.span {a})) (Ideal.mulQuotInjective (Ideal.span {a}) hb)
+          (Ideal.quotOfMulSurjective (Ideal.span {a})) (Ideal.quotMulExact (Ideal.span {a}))
+  simp only [Ring.ord, submodule_span_eq, ← this]
+  have lem : (({b} : Set R) • Ideal.span {a}) = Ideal.span {b * a} := by
+    simp [←Ideal.submodule_span_eq, Submodule.set_smul_span]
+  have : (({b} : Set R) • Ideal.span {a}) = b • Ideal.span {a} := Submodule.singleton_set_smul
+    (Ideal.span {a}) b
+  rw [this] at lem
+  rw [lem, mul_comm]
+
+open Classical in
+/--
+Monoid with zero homomorpihsm from a nontrivial commutative ring `R` to `ℕₘ₀`.
+
+Note that we cannot just use `fun x ↦ ord R x` without further assumptions on `R`.
+This is because if R is finite length, then ord R 0 will be some non top value,
+meaning in this case `0` will not be mapped to `⊤`.
+-/
+@[stacks 02MD]
+noncomputable
+def CommRing.ordMonoidWithZeroHom [Nontrivial R] : R →*₀ ℕₘ₀ where
+  toFun x := if x ∈ nonZeroDivisors R then Ring.ord R x else 0
+  map_zero' := by
+    simp [nonZeroDivisors, exists_ne]
+  map_one' := by
+    simp [nonZeroDivisors, Ring.ord_one]
+    rfl
+  map_mul' := by
+    intro x y
+    split_ifs
+    · exact CommRing.ord_mul R (by expose_names; exact h_2)
+    all_goals simp_all [mul_mem_nonZeroDivisors]
+
+/--
+Cast `ℕₘ₀` to `ℤₘ₀`.
+-/
+noncomputable
+def WithZero.castNatMulZeroMonoidWithZeroHom : ℕₘ₀ →*₀ ℤₘ₀ := WithZero.map' <|
+  (AddMonoidHom.toMultiplicative).toFun (Nat.castAddMonoidHom ℤ)
+
```

Comment:
This also shouldn't be here.

Reply:
The former. But the latter might also be true. The spelling should be `WithZero.map' (Nat.castAddMonoidHom ℤ).toMultiplicative` and it is actually not that longer compared to `WithZero.castNatMulZeroMonoidWithZeroHom`.